File: rajtszamok/valasz.pas
unit valasz;

interface
  Function induloszam: integer;
  Function kerdes(i,j: integer): integer;
  Procedure megoldas(i: integer);
  
implementation      
  const 
		Init:boolean=false;
		maxn=1024;
  var 
		n,db: integer;
    r: array[1..maxn] of integer;
//
procedure Vege(pont:integer; Uzi:string);
    begin
      writeln(pont);
      writeln(Uzi);
      halt;
    end;
    
  Function induloszam: integer;
    var i: integer;
  begin
    Init:=true;
    readln(n); db:=0;
    for i:=1 to n do read(r[i]);
    readln;
    induloszam:=n;
  end;
   
  Function kerdes(i,j: integer): integer;
  begin
  if not Init then Vege(0,'Protokoll hiba');
    inc(db);
    if (j>=0) and (j<=31) and (i>=1) and (i<=n) then kerdes:=(r[i] shr j) mod 2
    else Vege(0,'Protokoll hiba');
  end;
   
  Procedure megoldas(i: integer);
  begin
		if not Init then Vege(0,'Protokoll hiba');
    if i=r[n] then begin
			if db<=2*n-2 then begin
				Vege(3,'Helyes, kérdésszám <= 2*N-2');
			end else if db<=3*n-4 then begin
				Vege(2,'Helyes, kérdésszám <= 3*N-4');
			end else if db<=10*n then begin
				Vege(1,'Helyes, kérdésszám <= 10*N');
			end else begin
				Vege(0,'Helyes válasz, de a kérdésszám > 10*N');
			end;
    end else begin
			Vege(0,'Hibás válasz');
		end;
  end;
   
begin
end.

File: rajtszamok/valasz.cpp
#include <bits/stdc++.h>
#include "valasz.h"
#define maxN 10001
using namespace std;

int _r[maxN];
int _n, _db=0;
bool _Init = false;

void Vege(string pont, string s){
   cout<<pont<<endl;
   cout<<s<<endl;
   exit(0);
}
int induloszam(){
   cin>>_n;
   _db=0;
   for(int i=1; i<=_n; i++) {
      cin>>_r[i];
   }
   _Init=true;
   return _n;
}
void megoldas(int m){
   if(!_Init) Vege("0","Protokoll hiba");
   if(m==_r[_n]){
      if(_db<=2*_n-2){
         Vege("1","Helyes, kérdésszám <= 2*n-2");
      }else if(_db<=3*_n-4){
         Vege("0.4","Helyes, kérdésszám <= 3*n-4");
      }else if(_db<=10*_n){
         Vege("0.2","Helyes, kérdésszám <= 10*n");
      }else{
         Vege("0","Helyes, de a kérdésszám >10*n");
      }
   }else{
      Vege("0","Hibás megoldás");
   }
   return;
}

int kerdes(int i, int j){
   if(!_Init) Vege("0","Protokoll hiba");
   _db++;
   int v=0;
   if(i>=1 && i<_n && j>=0 && j<=31){
      v=(_r[i] >> j) %2;
      return v;
   }else{
      Vege("0","Protokoll hiba");
   }

   return v;
}

